src/main/kotlin/MinimaxPlayer.kt

```diff
@@ -0,0 +1,112 @@
+        private val squareFrom = model.board[turn.from]
+        private val squareFromFigure = squareFrom.figure?.copy()
+        private val squareTo = model.board[turn.to]
+        private val squareToFigure = squareTo.figure?.copy()
+        private val squareEaten = model.canMove(turn)
+        private val eatenFigure = squareEaten?.figure?.copy()
+        val whoMoves = model.whoMoves
+        private val eatingChecker = model.eatingChecker
+
+        fun reset(model: CheckersModel) {
+            squareEaten?.let {
+                it.figure = eatenFigure
+            }
+            squareFrom.figure = squareFromFigure
+            squareTo.figure = squareToFigure
+            model.whoMoves = whoMoves
+            model.gameState = GameState.PLAYING
+            model.eatingChecker = eatingChecker
+        }
+    }
+
+    protected val turnsByHash : MutableMap<Int, BaseTurn> = mutableMapOf()
+
+    //Function that performs the necessary moves
+    private fun minimaxRecursive(model : CheckersModel, depth: Int, maxDepth: Int) : Pair<Int, BaseTurn?> {
+        recCount++
+        val isTerminalStateResult = isTerminalState(model, depth, maxDepth)
+        if (isTerminalStateResult != null) {
+            return isTerminalStateResult to null
+        }
+        //Maybe, we already had this position in not so much depth, then we can use early result
+        val turns =/* if (turnsByHash[model.hashCode()] != null)
+            listOf(turnsByHash[model.hashCode()]!!)
+        else*/
+            model.possibleTurns()
+
+        var bestVal = -(Int.MAX_VALUE / 10)
+        var bestTurn = turns[0]
+        for (turn in turns) {
+            val retainer = ModelRetainer(model, turn)
+            try {
+                model.move(turn)
+            } catch (e : Exception) {
+                logger.log(Level.WARNING, turn.toString())
+                model.board.print()
+            }
+            val ans = minimaxRecursive(model, depth + if (model.whoMoves == retainer.whoMoves) 0 else 1, maxDepth).first *
+                    (if (model.whoMoves == retainer.whoMoves) 1 else -1)
+            retainer.reset(model)
+            if (depth == 0) {
+                logger.log(Level.INFO, turn.toString() + " : " + ans)
+            }
+            if (ans > bestVal) {
+                bestVal = ans
+                bestTurn = turn
+            }
+        }
+        if (depth < maxDepth / 2) {
+            turnsByHash[model.hashCode()] = bestTurn
+        }
+        return bestVal to bestTurn
+    }
+
+    protected fun isTerminalState(model: CheckersModel, depth: Int, maxDepth: Int): Int? {
+        if (model.gameState != GameState.PLAYING) {
+            var value = Int.MAX_VALUE / 5 + (100 - depth)
+            if (model.gameState.getColor() != model.whoMoves) {
+                value *= -1
+            }
+            return value
+        }
+        if (depth >= maxDepth && (!model.canEat() || depth >= maxDepth * 5)) {
```

Comment:
No, maxDepth means how deep our tree of miniMax could be. We need it because we can't look at all the positions, so if our tree is deeper, we will return

Reply:
So, how can it turn out that `depth >= maxDepth * 5`? If depth is incremented by one for each level of recursion and this condition is checked on each level? Isn't `depth >= maxDepth` enough?